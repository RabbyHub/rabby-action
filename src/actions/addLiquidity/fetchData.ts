import PQueue from 'p-queue';
import {
  AddLiquidityRequireData,
  FetchActionRequiredData,
  ParsedActionData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import BigNumber from 'bignumber.js';

export const fetchDataAddLiquidity: FetchActionRequiredData<{
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: any;
  token1: any;
}> = async (options, likeAction) => {
  const { actionData, apiProvider, sender, walletProvider, chainId } = options;

  const action =
    likeAction || (<ParsedActionData<'transaction'>>actionData).addLiquidity;

  if (!action || options.type === 'text') {
    return {};
  }

  const { receiver, exchange_rate, token0, token1 } = action;

  const fair_exchange_rate = new BigNumber(token0.price).div(token1.price);

  const diff = +new BigNumber(exchange_rate)
    .minus(fair_exchange_rate)
    .abs()
    .div(fair_exchange_rate)
    .times(100)
    .toFixed(4);

  const receiverInWallet = await walletProvider.hasAddress(receiver);

  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;
  if (!id) {
    return {};
  }

  const result: AddLiquidityRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: false,
    receiver,
    poolRate: exchange_rate,
    token0,
    token1,
    marketRate: fair_exchange_rate.toFixed(),
    diff,
  };
  const queue = new PQueue();

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;

      result.protocol = contractInfo.protocol;
    }
  });

  queue.add(async () => {
    const hasInteraction = await apiProvider.hasInteraction(
      sender,
      chainId,
      id
    );
    result.hasInteraction = hasInteraction.has_interaction;
  });

  await waitQueueFinished(queue);

  return result;
};
