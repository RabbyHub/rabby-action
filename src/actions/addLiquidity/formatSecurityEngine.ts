import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
