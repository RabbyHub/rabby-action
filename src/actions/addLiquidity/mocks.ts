export const preExecData = {
  abi: {
    func: 'mint',
    params: [
      [
        '0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22',
        '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913',
        10,
        -4970,
        -4960,
        2955759909,
        885199302,
        2940981110,
        880773306,
        '0xE2866F2291B5c5bafD2858B9247f93745108A523',
        1754366820,
        0,
      ],
    ],
  },
  abi_str:
    '{\n    "func": "mint",\n    "params": [\n        [\n            "0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22",\n            "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",\n            10,\n            -4970,\n            -4960,\n            2955759909,\n            885199302,\n            2940981110,\n            880773306,\n            "0xE2866F2291B5c5bafD2858B9247f93745108A523",\n            1754366820,\n            0\n        ]\n    ]\n}',
  balance_change: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 29320,
    gas_limit: 0,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'base',
    chain: 'base',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 3629.55,
    price_24h_change: 0.018385589869839923,
    credit_score: 46503412.60907772,
    is_verified: true,
    is_scam: false,
    is_suspicious: null,
    is_core: true,
    is_wallet: true,
    time_at: null,
    low_credit_score: false,
    amount: 1e-5,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
  },
  type_call: {
    action: 'mint',
    contract: '0x827922686190790b37229fd06084350e74485b72',
    contract_protocol_name: 'Aerodrome V3',
    contract_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/base_aerodrome/f02d753bc321dc8ba480f0424a686482.png',
  },
  trace_id: '663ae6c6c0ac4b1aa5651fadba316de2',
} as any;

export const parseTxData = {
  action: { type: 'cancel_tx' },
  contract_call: null,
  log_id: 441860918,
} as any;

export const txData = {
  chainId: 42161,
  from: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  to: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  value: '0x0',
  data: '0x',
  gas: '0x0',
  maxFeePerGas: '0xb71b00',
  maxPriorityFeePerGas: '0xb71b00',
  nonce: '0x4',
};
