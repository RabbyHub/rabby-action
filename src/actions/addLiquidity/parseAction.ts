import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
