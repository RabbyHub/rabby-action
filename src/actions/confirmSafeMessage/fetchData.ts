import { FetchActionRequiredData, MultiSigRequireData } from '../../types';

export const fetchDataConfirmSafeMessage: FetchActionRequiredData = async (
  options
) => {
  if (options.type !== 'typed_data') {
    return {};
  }
  const { actionData, apiProvider } = options;
  if (!actionData.confirmSafeMessage) {
    return {};
  }

  const result: MultiSigRequireData = {
    contract: null,
    id: actionData.confirmSafeMessage.multisig_id,
  };
  const { desc } = await apiProvider.addrDesc(
    actionData.confirmSafeMessage.multisig_id
  );
  if (desc.contract) {
    result.contract = desc.contract;
  }
  return result;
};
