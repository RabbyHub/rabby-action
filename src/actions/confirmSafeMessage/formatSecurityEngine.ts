import { FormatSecurityEngineContext } from '../../types';

export const formatSecurityEngineConfirmSafeMessage: FormatSecurityEngineContext =
  async () => {
    return {};
  };
