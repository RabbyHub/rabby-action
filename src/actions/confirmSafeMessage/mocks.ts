export const parseTxData = {
  action: {
    type: 'confirm_safe_message',
    data: {
      multisig_id: '0x89f4a8e75da4b985bae9798162073fe5c037e9dd',
    },
  },
  log_id: 10317007,
} as any;

export const txData = {
  types: {
    SafeMessage: [
      {
        name: 'message',
        type: 'bytes',
      },
    ],
    EIP712Domain: [
      {
        name: 'chainId',
        type: 'uint256',
      },
      {
        name: 'verifyingContract',
        type: 'address',
      },
    ],
  },
  primaryType: 'SafeMessage',
  domain: {
    chainId: '0x1',
    verifyingContract: '0x89f4a8e75da4b985bae9798162073fe5c037e9dd',
  },
  message: {
    message:
      '0x2bf01edd92e8a198c95b799831abdae4f93a2428674d06dcf39b4a3e80b38a09',
  },
};
