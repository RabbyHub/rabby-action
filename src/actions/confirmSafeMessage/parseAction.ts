import { ConfirmSafeMessageActions } from '@rabby-wallet/rabby-api/dist/types';
import { PartialParseAction } from '../../types';

export const parseActionConfirmSafeMessage: PartialParseAction<'typed_data'> = (
  options
) => {
  const { data } = options;

  if (data?.type !== 'confirm_safe_message') {
    return {};
  }
  return {
    confirmSafeMessage: data.data as ConfirmSafeMessageActions,
  };
};
