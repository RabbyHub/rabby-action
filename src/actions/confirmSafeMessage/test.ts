import {
  apiProvider,
  ETH_CHAIN_ID,
  formatProvider,
  ORIGIN,
  SENDER,
  walletProvider,
} from '../../../__mocks__';
import { fetchDataConfirmSafeMessage } from './fetchData';
import { formatSecurityEngineConfirmSafeMessage } from './formatSecurityEngine';
import { parseTxData, txData } from './mocks';
import { parseActionConfirmSafeMessage } from './parseAction';
import { parseTypedDataAction } from '../../utils/parseTypedDataAction';
import {
  parseAction,
  fetchActionRequiredData,
  formatSecurityEngineContext,
} from '../..';

test.each([
  [
    parseTypedDataAction(parseActionConfirmSafeMessage),
    fetchDataConfirmSafeMessage,
    formatSecurityEngineConfirmSafeMessage,
  ],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('Confirm Safe Message', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'typed_data',
    data: parseTxData['action'],
    typedData: txData,
    sender: SENDER,
  });
  expect(actionData).toMatchSnapshot('parseActionConfirmSafeMessage');

  const requireData = await _fetchData({
    type: 'typed_data',
    actionData,
    chainId: ETH_CHAIN_ID,
    walletProvider,
    apiProvider,
    sender: SENDER,
  });
  expect(requireData).toMatchSnapshot('fetchDataConfirmSafeMessage');

  const ctx = await _format({
    type: 'typed_data',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
    origin: ORIGIN,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineConfirmSafeMessage');
});
