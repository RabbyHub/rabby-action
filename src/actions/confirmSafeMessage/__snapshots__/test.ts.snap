// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Confirm Safe Message: fetchDataConfirmSafeMessage 1`] = `
{
  "contract": {},
  "id": "0x89f4a8e75da4b985bae9798162073fe5c037e9dd",
}
`;

exports[`Confirm Safe Message: fetchDataConfirmSafeMessage 2`] = `
{
  "bornAt": 1591388241,
  "hasInteraction": true,
  "id": "0x89f4a8e75da4b985bae9798162073fe5c037e9dd",
  "protocol": {
    "id": "uniswap2",
    "logo_url": "https://static.debank.com/image/project/logo_url/uniswap2/87a541b3b83b041c8d12119e5a0d19f0.png",
    "name": "Uniswap V2",
  },
  "rank": 65,
  "receiverInWallet": false,
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
  "unexpectedAddr": null,
}
`;

exports[`Confirm Safe Message: formatSecurityEngineConfirmSafeMessage 1`] = `{}`;

exports[`Confirm Safe Message: formatSecurityEngineConfirmSafeMessage 2`] = `
{
  "contractCall": {
    "chainId": "eth",
    "id": "0x89f4a8e75da4b985bae9798162073fe5c037e9dd",
  },
}
`;

exports[`Confirm Safe Message: parseActionConfirmSafeMessage 1`] = `
{
  "actionType": "confirm_safe_message",
  "brand": undefined,
  "chainId": "0x1",
  "confirmSafeMessage": {
    "multisig_id": "0x89f4a8e75da4b985bae9798162073fe5c037e9dd",
  },
  "contractId": "0x89f4a8e75da4b985bae9798162073fe5c037e9dd",
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;

exports[`Confirm Safe Message: parseActionConfirmSafeMessage 2`] = `
{
  "actionType": "contractCall",
  "brand": undefined,
  "chainId": "0x1",
  "contractCall": {},
  "contractId": "0x89f4a8e75da4b985bae9798162073fe5c037e9dd",
  "sender": "0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85",
}
`;
